### PR TITLE
Add docs for the haproxy config manager to reduce reloads.

### PR DESCRIPTION
**Pull request by ramr**
i am still debating on how to add the rest of the details to this - need to explain blueprint routes and usage etc. I'll probably add those details on another PR as I suspect it will need more clarifications (look at this as simple usage and that one for more advanced usage/control).
@bfallonf  PTAL Thx

@ironcladlou @lihongan  fyi

### CODE REVIEW

**Comment by bfallonf** [comment]
Thanks @ramr ! Does this apply to an existing BZ or trello card? That's helpful to determine the timeframe to work on and also which version it applies to.

**Comment by ramr** [comment]
@bfallonf my bad - i linked on the trello card but not here.
It is for 3.11 and the associated card is: https://trello.com/c/W21FD0v0/20-13-routerscale-implement-dynamic-changes-to-the-router

**Comment by lihongan** [comment]
I guess the `ROUTER_MAX_DYNAMIC_SERVERS` and `COMMIT_INTERVAL` can be added to this PR, if you plan to add blueprint to another one.

**Comment by ramr** [comment]
@lihongan  good catch thx - will do.

File: install_config/router/default_haproxy_router.adoc [hunk 1]
```diff
@@ -1487,6 +1487,94 @@ add routes across the namespaces.
...
+API with a pool of pre-allocated routes and backend servers to dynamically
+achieve the same results without requiring a haproxy process reload.
+In the event the dynamic configuration manager is unable to configure haproxy
+dynamically, it forces a configuration re-write and a haproxy process reload.
+This can occur in cases like if a newly added route contains custom annotations
+that would cause a drift from the standard configuration (custom timeouts) or
+if the route requires custom TLS configuration.
+The pre-allocated pool of routes is created using route blueprints - the default
+set of blueprints supports unsecured routes, edge secured routes (without any
+custom TLS configuration) and passthrough routes.
```
Comment (lihongan) [inline install_config/router/default_haproxy_router.adoc]: Can users `get` or `edit` the default set of blueprints? Does `ROUTER_BLUEPRINT_ROUTE_POOL_SIZE` also control the default blueprints? I mean if each route of default blueprints has 10 pools by default?

File: install_config/router/default_haproxy_router.adoc [hunk 2]
```diff
@@ -1487,6 +1487,94 @@ add routes across the namespaces.
...
+`re-encypt` routes require custom TLS configuration information and so are not
+supported out of the box. However that said, the blueprints that the dynamic
+configuration manager uses can be extended by setting the `ROUTER_BLUEPRINT_ROUTE_NAMESPACE`
+and optionally the `ROUTER_BLUEPRINT_ROUTE_LABELS` environment variable values.
```
Comment (lihongan) [inline install_config/router/default_haproxy_router.adoc]: Can users configure multiple `ROUTE_NAMESPACE` and `ROUTE_LABELS` for custom blueprints?

File: install_config/router/default_haproxy_router.adoc (hunk 1 above)
Comment (ramr) [reply install_config/router/default_haproxy_router.adoc]: Well it all depends on who `owns` the namespace for the blueprints and what access is allowed?
In the normal use case, the admin running the router would `own` that namespace and so an end user (one creating routes) would *_NOT_* have that access.

As re: the `ROUTER_BLUEPRINT_ROUTE_POOL_SIZE`, yes it controls the default blueprints as well as any custom ones the admin defines. 
And each one of those blueprints (default or custom)  ... for example blueprint-a  would have 10 (default) pre-defined backends - blueprint-a-1 through blueprint-a-10 - inside the haproxy config.

File: install_config/router/default_haproxy_router.adoc (hunk 2 above)
Comment (ramr) [reply install_config/router/default_haproxy_router.adoc]: It depends on permissions to that project/namespace `ROUTE_NAMESPACE`. 
The general use case would be no - as the admin owns the `ROUTE_NAMESPACE` so only the admin can add a `LABEL` to those blueprint routes. But an admin can always grant access to another user to modify that namespace and/or the routes.

**Comment by vikram-redhat** [comment]
@ramr are there any more updates coming for this PR?

**Comment by ramr** [comment]
@vikram-redhat no more updates here - it is good from my end. Thx.
btw, build is failing on an unrelated change ... fails on `topics/increasing_docker_storage.adoc`.

**Comment by bfallonf** [comment]
Thanks @ramr . With the broken build, I've pulled this PR and I'll make some changes on my end. I'll create a new PR and bug you with any questions.

**Comment by bfallonf** [comment]
I've created #11712 . @ramr , can I please ask for a review? I've made the topic more into a procedure, and I hope that I haven't broken anything. Thanks.

I'll close this PR.

**Pull request closed without merging**